src/ProtobufServer/threadpool.rs

```diff
@@ -0,0 +1,125 @@
+use crossbeam_channel::{unbounded, Receiver, Sender};
+use futures::Future;
+use num_cpus;
+use std::sync::{Arc, Mutex};
+use std::task::{Poll, Waker};
+use std::thread::{self, JoinHandle};
+use tonic::Status;
+
+use crate::DataReader::component_loader::ModelCache;
+use crate::ProtobufServer::services::{QueryRequest, QueryResponse};
+
+use crate::ProtobufServer::ConcreteEcdarBackend;
+
+type ThreadPoolResponse = Result<QueryResponse, Status>;
+
+/// A construct that uses a fixed amount of threads to do work in parallel.
+#[derive(Debug)]
+pub struct ThreadPool {
+    sender: Option<Sender<Context>>,
+    threads: Vec<JoinHandle<()>>,
+}
+
+impl ThreadPool {
+    /// Create a new thread pool.
+    ///
+    /// # Arguments
+    ///
+    /// * `num_threads` - The amount of threads in the thread pool.
+    pub fn new(num_threads: usize) -> Self {
+        let (sender, receiver): (Sender<Context>, Receiver<Context>) = unbounded();
+        let cache = ModelCache::default();
+
+        let threads = (0..num_threads)
+            .map(|_| {
+                let thread_receiver = receiver.clone();
+                let thread_cache = cache.clone();
+                thread::spawn(move || {
+                    for mut context in thread_receiver {
+                        let query_response = ConcreteEcdarBackend::handle_send_query(
+                            context.query_request,
+                            thread_cache.clone(),
+                        );
+                        context.future.complete(query_response);
+                    }
+                })
+            })
+            .collect();
+
+        ThreadPool {
+            sender: Some(sender),
+            threads,
+        }
+    }
+
+    /// Enqueue a query request. Returns a future that can be awaited to get a `QueryResponse`.
+    ///
+    /// # Arguments
+    ///
+    /// * `query_request` - the query request to enqueue.
+    pub fn enqueue(&self, query_request: QueryRequest) -> ThreadPoolFuture {
+        let future = ThreadPoolFuture::default();
+        let context = Context {
+            future: future.clone(),
+            query_request,
+        };
+        self.sender.as_ref().unwrap().send(context).unwrap();
+        future
+    }
+}
+
+impl Default for ThreadPool {
+    fn default() -> Self {
+        Self::new(num_cpus::get())
+    }
+}
+
+impl Drop for ThreadPool {
+    fn drop(&mut self) {
+        // Drop the sender to signal the threads to quit
+        self.sender = None;
+
+        loop {
+            match self.threads.pop() {
+                None => break,
+                Some(thread) => thread.join().unwrap(),
+            }
+        }
```

Comment:
Is it possible to omit this loop?
```rust
for thread in self.threads.iter() {
    thread.join().unwrap();
}
```

Reply:
No join does not work on references and we cant move it out of threads

Reply:
fair